File: examples/python/src/circle.py
"""Circle shape implementation."""

from __future__ import annotations

import math

from shape import Color, Shape, register_shape
from vector3 import Vector3


@register_shape
class Circle(Shape):
    def __init__(
        self,
        center: Vector3[float],
        radius: float,
        color: Color = Color.BLUE,
    ) -> None:
        super().__init__("Circle", color)
        self._center = center
        self._radius = radius

    @property
    def radius(self) -> float:
        return self._radius

    @property
    def center(self) -> Vector3[float]:
        return self._center

    def area(self) -> float:
        return math.pi * self._radius**2

    def perimeter(self) -> float:
        return 2.0 * math.pi * self._radius

    def centroid(self) -> Vector3[float]:
        return self._center

    def describe(self) -> str:
        base = super().describe()
        return f"{base}\n  radius={self._radius}  center={self._center}"

File: examples/javascript/src/circle.js
import { Color, Shape, registerShape } from "./shape.js";
import { Vector3 } from "./vector3.js";

export class Circle extends Shape {
  #center;
  #radius;

  constructor(center = new Vector3(), radius = 1, color = Color.BLUE) {
    super("Circle", color);
    this.#center = center;
    this.#radius = radius;
  }

  get radius() { return this.#radius; }
  get center() { return this.#center; }

  area()      { return Math.PI * this.#radius ** 2; }
  perimeter() { return 2 * Math.PI * this.#radius; }
  centroid()  { return this.#center; }

  describe() {
    return `${super.describe()}\n  radius=${this.#radius}  center=${this.#center}`;
  }
}

registerShape(Circle);
